core/trino-main/src/main/java/io/trino/execution/scheduler/FaultTolerantQueryScheduler.java

```diff
@@ -268,45 +265,45 @@ private Scheduler createScheduler()
                         systemTicker(),
                         exchange,
                         sinkPartitioningScheme,
+                        sourceSchedulers.buildOrThrow(),
                         sourceExchanges.buildOrThrow(),
                         sourcePartitioningScheme,
                         remainingTaskRetryAttemptsOverall,
                         taskRetryAttemptsPerTask,
                         maxTasksWaitingForNodePerStage,
                         dynamicFilterService);
 
-                schedulers.add(scheduler);
-            }
+                schedulers.put(fragment.getId(), scheduler);
 
-            if (!stagesInReverseTopologicalOrder.isEmpty()) {
-                verify(!outputStages.isEmpty(), "coordinatorConsumedExchanges is empty");
-                List<ListenableFuture<List<ExchangeSourceHandle>>> futures = outputStages.stream()
-                        .map(Exchange::getSourceHandles)
-                        .map(Exchanges::getAllSourceHandles)
-                        .collect(toImmutableList());
-                ListenableFuture<List<List<ExchangeSourceHandle>>> allFuture = Futures.allAsList(futures);
-                addSuccessCallback(allFuture, result -> {
-                    List<ExchangeSourceHandle> handles = result.stream()
-                            .flatMap(List::stream)
-                            .collect(toImmutableList());
-                    ImmutableList.Builder<ExchangeInput> inputs = ImmutableList.builder();
-                    if (!handles.isEmpty()) {
-                        inputs.add(new SpoolingExchangeInput(handles));
-                    }
-                    queryStateMachine.updateInputsForQueryResults(inputs.build(), true);
-                });
-                addExceptionCallback(allFuture, queryStateMachine::transitionToFailed);
+                // root stage
```

Comment:
what did change here?

Reply:
Let me better extract it into a separate commit